### DISCUSSION TITLE
GitHub REST API CORS error when trying to download public repository git archive

### DISCUSSION THREAD

**Discussion by Crystal-Spider** [community/community#106849] [Apps, API and Webhooks] [answered]
### Select Topic Area

Bug

### Body

Not sure if this is an actual bug or I'm just misunderstanding something.  
As suggested, I reference [my comment](https://github.com/github/docs/issues/24706#issuecomment-1933107348) in [this closed issue](https://github.com/github/docs/issues/24706).

I also copied it down here for ease of reading.

---

Hi, sorry for commenting on a closed issue, but I have a problem that seems relevant to this.

I'm trying to build an Angular application hosted on my GitHub pages with a custom verified domain.  
This is my code to call GitHub API and get the zip archive of a (public) repository of mine:
```typescript
  public getTemplate() {
    return this.http.get(
      'https://api.github.com/repos/crystal-nest/cobweb-mod-template/zipball/1.20.4',
      {
        headers: {
          'X-GitHub-Api-Version': '2022-11-28'
        }
      }
    );
  }
```
I know the URL is correct because if I paste it in my browser the download starts.  
However, when the code gets executed I get the following error:

> Access to XMLHttpRequest at 'https://codeload.github.com/Crystal-Nest/cobweb-mod-template/legacy.zip/refs/heads/1.20.4' (redirected from 'https://api.github.com/repos/crystal-nest/cobweb-mod-template/zipball/1.20.4') from origin 'https://crystalnest.it' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

I then tried with [JSONP requests](https://docs.github.com/en/rest/using-the-rest-api/using-cors-and-jsonp-to-make-cross-origin-requests?apiVersion=2022-11-28#json-p-callbacks):
```typescript
return this.http.jsonp('https://api.github.com/repos/crystal-nest/cobweb-mod-template/zipball/1.20.4', 'callback');
```
But this time the error I received was:
```json
{
    "headers": {
        "normalizedNames": {},
        "lazyUpdate": null,
        "headers": {}
    },
    "status": 0,
    "statusText": "JSONP Error",
    "url": "https://api.github.com/repos/crystal-nest/cobweb-mod-template/zipball/1.20.4?callback=ng_jsonp_callback_0",
    "ok": false,
    "name": "HttpErrorResponse",
    "message": "Http failure response for https://api.github.com/repos/crystal-nest/cobweb-mod-template/zipball/1.20.4?callback=ng_jsonp_callback_0: 0 JSONP Error",
    "error": {
        "message": "JSONP injected script did not invoke callback.",
        "stack": "Error: JSONP injected script did not invoke callback.\n    stacktrace omitted for the sake of brevity"
    }
}
```

My code can be found [here](https://github.com/Crystal-Nest/crystalnest.it), and [here](https://crystalnest.it/generator) there's a live example (click the download button and ignore everything else, the website is very WIP at the moment).

Side note: even if the request worked and I was able to get the file as an `'arraybuffer'` (the data I need to process the file with JSZip), it's a pain that the zip name and its root directory have a code appended at the end (that I can't understand what it is and thus I can't predict what it's going to be). Ideally, I'd like to use this URL `https://github.com/crystal-nest/cobweb-mod-template/archive/refs/heads/1.20.4.zip` because it gives me a name easy to predict and thus easy to process.

Let me know whether it's better I open a new issue, thanks in advance.

**Comment by gr2m** [comment, marked as answer]
Unfortunately this is a problem with CORS settings for adjacent domains. This came up years ago when I was working on @octokit, I summarized my own wishlist for GitHub's API here: https://github.com/gr2m/github-api-wishlist/tree/master/wishlist/cors-for-adjacent-domains

The last I remember is that CORS setting would not be added to these secondary domains due to security concerns. I suggested we introduce a new media type so that the redirect URL can be retrieved without CORS (https://github.com/gr2m/github-api-wishlist/tree/master/wishlist/link-media-type), but that work was not prioritized so far.

I fear your best option is to create a proxy at this point.

**Reply by Crystal-Spider** [reply]
Would anything change if I used the github.io domain rather than my custom domain?

**Reply by gr2m** [reply]
no. CORS settings need to be set on the API domains, the request origin doesn't matter. The CORS headers are missing entirely

**Reply by Crystal-Spider** [reply]
I hope someday the new media type will come true, I don't really like having to use a proxy to avoid CORS. But at least until then, now I can move forward. Thanks!

**Reply by drwpow-figma** [reply]
Noooo 🙃 I just hit this

**Comment by Crystal-Spider** [comment]
As pointed out on the accepted answer, the only current way to do it is by employing a proxy.
In my case, I simply used [this free and public proxy](https://corsproxy.io/).

**Comment by marcanii** [comment]
responsive #2

**Comment by gopro2027** [comment]
corsproxy io no longer works (Website even downloads a weird streaming file when you go it it? Odd...)

Please create a [worker on Cloudflare](https://workers.cloudflare.com/) and use this code as a proxy (Working example for downloading assets from releases by tag name). Cloudflare is free up to 100,000 requests per day.

```Javascript
/**
 * Reverse proxy for GitHub releases
 * Usage: https://WORKERNAME.CLOUDFLAREUSERNAME.workers.dev/?user=${user}&repo=${repo}&tag_name=${tag_name}&file_name=${file_name}
 */
export default {
  async fetch(request, env, ctx) {
      const url = new URL(request.url);
      const user = url.searchParams.get('user');
      const repo = url.searchParams.get('repo');
      const tag_name = url.searchParams.get('tag_name');
      const file_name = url.searchParams.get('file_name');
      
      if (!user || !repo || !tag_name || !file_name) {
        return new Response('Missing required parameters: user or repo or tag_name or file_name', {status: 400, statusText: 'Bad Request'});
      }

      return new Response(
        await (await fetch(`https://github.com/${user}/${repo}/releases/download/${tag_name}/${file_name}`)).arrayBuffer(),
        {
          headers: {
            "Access-Control-Allow-Headers": request.headers.get("Access-Control-Request-Headers"),
            "Access-Control-Allow-Methods": "GET,HEAD,POST,OPTIONS",
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Max-Age": "300",
            "content-type": "application/octet-stream",
          }
        }
      );
  },
};
```
